mongocrypt-sys/src/test/example_state_machine.rs

```diff
@@ -0,0 +1,288 @@
+                    assert!(mongocrypt_kms_ctx_message(kms, output));
+                    with_binary_as_slice(output, |slice| {
+                        println!(
+                            "sending the following to kms:\n{:?}",
+                            str::from_utf8(slice).unwrap()
+                        );
+                    });
+                    mongocrypt_binary_destroy(output);
+                    let input = BinaryBuffer::read_http("src/test/testdata/kms-decrypt-reply.txt");
+                    println!(
+                        "mocking reply from file:\n{:?}",
+                        str::from_utf8(&input.bytes).unwrap()
+                    );
+                    assert!(mongocrypt_kms_ctx_feed(kms, input.binary));
+                    assert_eq!(0, mongocrypt_kms_ctx_bytes_needed(kms));
+                }
+                mongocrypt_ctx_kms_done(ctx);
+            }
+            mongocrypt_ctx_state_t_MONGOCRYPT_CTX_READY => {
+                let output = mongocrypt_binary_new();
+                assert!(mongocrypt_ctx_finalize(ctx, output));
+                result = doc_from_binary(output);
+                mongocrypt_binary_destroy(output);
+            }
+            mongocrypt_ctx_state_t_MONGOCRYPT_CTX_DONE => {
+                break;
+            }
+            mongocrypt_ctx_state_t_MONGOCRYPT_CTX_ERROR => {
+                let status = mongocrypt_status_new();
+                mongocrypt_ctx_status(ctx, status);
+                let message = CStr::from_ptr(mongocrypt_status_message(status, ptr::null_mut()))
+                    .to_str()
+                    .unwrap();
+                panic!("got error: {}", message);
+            }
+            state => panic!("unhandled state {:?}", state),
+        }
+    }
+
+    return result;
+}
+
+unsafe extern "C" fn log_to_stderr(
+    level: mongocrypt_log_level_t,
+    message: *const ::std::os::raw::c_char,
+    _message_len: u32,
+    _ctx: *mut ::std::os::raw::c_void,
+) {
+    #[allow(non_upper_case_globals)]
+    let level_str = match level {
+        mongocrypt_log_level_t_MONGOCRYPT_LOG_LEVEL_ERROR => "ERROR",
+        mongocrypt_log_level_t_MONGOCRYPT_LOG_LEVEL_WARNING => "WARNING",
+        mongocrypt_log_level_t_MONGOCRYPT_LOG_LEVEL_INFO => "INFO",
+        mongocrypt_log_level_t_MONGOCRYPT_LOG_LEVEL_TRACE => "TRACE",
+        _ => "UNKNOWN",
+    };
+    eprintln!("{}{}", level_str, CStr::from_ptr(message).to_str().unwrap());
+}
+
+fn cs(bytes: &'static [u8]) -> *const i8 {
+    CStr::from_bytes_with_nul(bytes).unwrap().as_ptr()
+}
+
+unsafe fn init_test_crypt() -> *mut mongocrypt_t {
+    let crypt = mongocrypt_new();
+    mongocrypt_setopt_kms_provider_aws(crypt, cs(b"example\0"), -1, cs(b"example\0"), -1);
+    mongocrypt_setopt_log_handler(crypt, Some(log_to_stderr), ptr::null_mut());
+    if !mongocrypt_init(crypt) {
+        panic!("failed to initialize");
+    }
+    crypt
+}
+
+#[test]
+fn encryption_decryption() {
+    unsafe {
```

Comment:
general curiosity as I haven't worked with `unsafe` code -- is there a difference between a function with an `unsafe` block and an `unsafe fn`? could these methods here that are entirely within an `unsafe` block be `unsafe fn`s?

Reply:
An `unsafe fn` can only be called from within another `unsafe` block/fn.  Morally, an `unsafe fn` is saying "upholding the safety invariants required for this function is on the caller" vs. a `fn` with an internal `unsafe` block is "this fn does unsafe things internally but callers don't need to worry about that".

  Sadly, tests aren't allowed to be `unsafe` or I would have done that here and saved the nesting.